Streaming Diarization Improvements

SGD2718 opened the pull request:
- Fixed `SegmentProcessor.getSegments` internal access level
- Added audio conversion method for `CMSampleBuffer`
- Added a `startTime` argument to `DiarizerManager.performCompleteDiarization` to offset segment timestamps.
- Added `AudioStream` struct to simplify timestamp calculations for a streamed/online audio input source with a lot of overlap

Alex-Wengg commented on Documentation/SpeakerDiarization.md:
```diff
@@ -1,3 +1,4 @@
+
 # Speaker Diarization
 
 Real-time speaker diarization for iOS and macOS, answering "who spoke when" in audio streams.
```
how real time we talking, can i do a 1 second chunking or what ? . a general subjective guidline. generally 0-5 seconds is considered "real-time" , ~ 10 seconds chunk is considered "near real time"

Alex-Wengg commented on Documentation/SpeakerDiarization.md:
```diff
@@ -255,35 +256,24 @@ Process audio in chunks for real-time applications:
-        }
+var stream = AudioStream(
+    chunkDuration: 5.0, // 10.0 for best accuracy. 3.0 for lowest
+    strideDuration: 2.0, // Duration between successive chunk starts
```
this line isn't very clear , can we try another variable instead of  strideDuration

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -62,18 +86,12 @@ public final class EmbeddingExtractor {
                 continue
             }
 
```
why is this removed

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -106,28 +124,60 @@ public final class EmbeddingExtractor {
         return embeddings
     }
 
+    private func fillWaveformBuffer<C>(
```
not very intuitive , maybe paddingAudio.

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -106,28 +124,60 @@ public final class EmbeddingExtractor {
+        }
+    }
+
     private func fillMaskBufferOptimized(
```
fillWaveformBuffer vs fillMaskBufferOptimized ?

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Segmentation/SegmentationProcessor.swift:
```diff
@@ -11,6 +11,81 @@ public struct SegmentationProcessor {
 
+    /// Repeat audio to fill the expected chunk length for downstream models.
+    ///
+    /// This keeps timestamps stable because only the audio is repeated; segment timing
```
these comments can be worded better

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -57,6 +58,46 @@ final public class AudioConverter {
         return try resampleBuffer(buffer)
     }
 
+    /// Convert a CMSampleBuffer to 16kHz mono Float32 samples
```
i feel like this is already done before , deja vu . are you able to verify

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -292,6 +336,12 @@ public enum AudioConverterError: LocalizedError {
         case .conversionFailed(let error):
             return "Audio conversion failed: \(error?.localizedDescription ?? "Unknown error")"
+        case .sampleBufferFormatMissing:
+            return "Sample buffer is missing a valid audio format description"
```
why would we need this ??

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -292,6 +336,12 @@ public enum AudioConverterError: LocalizedError {
+            return "Sample buffer is missing a valid audio format description"
+        case .frameCountOverflow:
+            return "Sample buffer contains more frames than supported"
+        case .sampleBufferCopyFailed(let status):
```
what could cause this error though

Alex-Wengg reviewed (commented):
first impression this looks pretty decent, i will review more in detail after 


are you able to add a full diarizer benchmark for near real time to benchmark.md ?

SGD2718 commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -51,6 +52,29 @@ public final class EmbeddingExtractor {
         var embeddings: [[Float]] = []
 
+        let audioSampleCount = audio.distance(from: audio.startIndex, to: audio.endIndex)
+        let effectiveSampleCount = min(audioSampleCount, waveformSampleCount)
```
The changes in the embeddingExtractor were made with Codex. I figured that they worked since the benchmark didn't indicate a noticeable slowdown, while the accuracy did improve.

SGD2718 commented on Documentation/SpeakerDiarization.md:
```diff
@@ -322,18 +312,30 @@ import AVFoundation
+            strideDuration: 3.0,
+            atTime: 0.0,
+            alignment: .frontAligned,
+            processGaps: Bool = false
```
The idea is that if the user pauses the app and the clock continues moving forward, then it will simply skip the long gap of silence when it resumes, instead of processing the silence. Although, in hindsight, I guess that could also be accomplished by leaving the time argument for the ScoreStream.write function blank.

SGD2718 commented on Documentation/SpeakerDiarization.md:
```diff
@@ -255,35 +256,24 @@ Process audio in chunks for real-time applications:
+var stream = AudioStream(
+    chunkDuration: 5.0, // 10.0 for best accuracy. 3.0 for lowest
+    strideDuration: 2.0, // Duration between successive chunk starts
+    atTime: 0.0, // Audio stream start time
```
The atTime argument is there so the stream may be aligned relative to a global clock instead of the start of the audio.

SGD2718 commented:
what happened to the ASR and VAD benchmarks i never even touched those

SGD2718 commented:
Summary of changes:
- Added `AudioStream` struct to facilitate timestamping for real-time audio streams with overlapping chunks
- `DiarizerManager.performCompleteDiarization()` now has a `startTime` parameter to remove the need for manual timestamp adjustments while streaming
- `EmbeddingExtractor.getEmbeddings` now loops the audio signal for chunks shorter than 10s
- Changed `SegmentationProcessor.getSegments()` access level from `internal` to a `public`
- Updated `DiarizationManager.md` to show a real-time diarization implementation using the AudioStream class instead of manual timestamping.
- Added missing command for real-time diarization benchmarks in `Benchmark.md`
- `ANEMemoryOptimizer.optimizedCopy` now performs a move for sources of type `ContiguousArray<Float>`
- The `process` CLI command now has arguments for `chunk-seconds` and `overlap-seconds`.
- The `diarization-benchmark` command now uses an `autoreleasepool` to avoid crashing when there are a lot of chunks

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -35,6 +35,7 @@ public final class EmbeddingExtractor {
         }
 
         let waveformShape = [3, 160_000] as [NSNumber]
+        let waveformSampleCount = waveformShape.last?.intValue ?? 160_000
```
why this tertiary specifically

SGD2718 replied:
Residue from Codex that i missed

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -106,8 +112,54 @@ public final class EmbeddingExtractor {
     }
 
+    /// Fill the waveform buffer with loop (repeat) padding
+    private func fillWaveformBufferOptimized<C>(
```
this feels convoluted, can you see it can be simplified or compacted

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -119,15 +171,35 @@ public final class EmbeddingExtractor {
+            )
+        }
+
+        guard numMasksUsed < maskCount && numMasksUsed > 0 else {
```
is this necessary

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -292,6 +336,12 @@ public enum AudioConverterError: LocalizedError {
+        case .frameCountOverflow:
+            return "Sample buffer contains more frames than supported"
+        case .sampleBufferCopyFailed(let status):
+            return "Failed to copy samples from CMSampleBuffer (status: \(status))"
```
what edge case would cause this error to pop up

SGD2718 replied:
For sampleBufferCopyFailed:
```swift
// AAC format description
var asbd = AudioStreamBasicDescription(
    mSampleRate: 44100,
    mFormatID: kAudioFormatLinearPCM,
    mFormatFlags: kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked,
    mBytesPerPacket: 2,
    mFramesPerPacket: 1,
    mBytesPerFrame: 2,
    mChannelsPerFrame: 1,
    mBitsPerChannel: 16,
    mReserved: 0
)

let asbdPtr = withUnsafePointer(to: &asbd) { $0 }

var desc: CMAudioFormatDescription?
CMAudioFormatDescriptionCreate(
    allocator: nil,
    asbd: asbdPtr,
    layoutSize: 0,
    layout: nil,
    magicCookieSize: 0,
    magicCookie: nil,
    extensions: nil,
    formatDescriptionOut: &desc
)

// Allocate a block buffer *smaller* than what PCM needs
var blockBuffer: CMBlockBuffer?
CMBlockBufferCreateWithMemoryBlock(
    allocator: nil,
    memoryBlock: nil,
    blockLength: 1,   // ❌ needs at least 2 bytes for 1 sample
    blockAllocator: nil,
    customBlockSource: nil,
    offsetToData: 0,
    dataLength: 1,
    flags: 0,
    blockBufferOut: &blockBuffer
)

var sampleBuffer: CMSampleBuffer?
CMSampleBufferCreate(
    allocator: nil,
    dataBuffer: blockBuffer,
    dataReady: true,
    makeDataReadyCallback: nil,
    refcon: nil,
    formatDescription: desc!,
    sampleCount: 1,
    sampleTimingEntryCount: 0,
    sampleTimingArray: nil,
    sampleSizeEntryCount: 0,
    sampleSizeArray: nil,
    sampleBufferOut: &sampleBuffer
)

guard let sampleBuffer else {
    return
}
let converter = AudioConverter()

do {
    _ = try converter.resampleSampleBuffer(sampleBuffer)
} catch {
    print(error)
}
```

This will print 

```
sampleBufferCopyFailed(-12704)
```

SGD2718 replied:
As for why anyone in their right mind would do this, I don't know.

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioStream.swift:
```diff
@@ -0,0 +1,387 @@
+    /// - Warning: Samples may be skipped if the time jumps forward significantly.
+    public mutating func write<C>(from source: C, atTime time: TimeInterval? = nil) throws
+    where C: Collection & Sequence, C.Element == Float {
+        if let array = source as? [Float] {
```
do these type checkings really help ? its an internal library so its very unlikely for the data types to be altered

Alex-Wengg commented on Tests/FluidAudioTests/AudioStreamTests.swift:
i believe we have a ASR STT streaming algo unit tests you can take inspirations from , i think the biggest issue we faced was related to timestamp syncing

SGD2718 replied:
Added some test cases to ensure timestamps work

SGD2718 commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -57,6 +58,46 @@ final public class AudioConverter {
         return try resampleBuffer(buffer)
     }
 
+    /// Convert a CMSampleBuffer to 16kHz mono Float32 samples
```
I think you're getting it mixed up with the one that converts the AVAudioPCMBuffer. Converting from a CMSampleBuffer is a bit different and is used when the audio stream comes from a live video feed, which has both microphone and video data.

SGD2718 commented on Sources/FluidAudio/Shared/AudioStream.swift:
```diff
@@ -0,0 +1,387 @@
+        sampleRate: Double = 16_000,
+        bufferCapacitySeconds: TimeInterval? = nil
+    ) throws {
+        guard chunkDuration >= 1 / sampleRate else {
```
I have changed this to be more clear by instead making the guard statement check if `chunkSize > 0`

Alex-Wengg commented on Sources/FluidAudio/Diarizer/Extraction/EmbeddingExtractor.swift:
```diff
@@ -118,7 +160,9 @@ public final class EmbeddingExtractor {
 
         // Copy speaker mask to first slot using optimized memory copy
-        let maskCount = masks[speakerIndex].count
+        let requiredMasks = masks[speakerIndex].count
```
shouldn't this be the other way , 

his is backwards. You're setting maskCount (current size) to numMasksUsed (which is already the calculated target size), then trying to pad up to masks[speakerIndex].count. The logic should be:
```
let requiredMasks = numMasksUsed  // Target size to pad to
var maskCount = min(masks[speakerIndex].count, numMasksUsed)  // Current size
```

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioConverter.swift:
```diff
@@ -292,6 +336,12 @@ public enum AudioConverterError: LocalizedError {
         case .conversionFailed(let error):
             return "Audio conversion failed: \(error?.localizedDescription ?? "Unknown error")"
+        case .sampleBufferFormatMissing:
+            return "Sample buffer is missing a valid audio format description. Make sure it's not a video sample."
```
`            return "Sample buffer is missing a valid audio format description. Make sure it's not a video sample." 
            `
can we make this more formal , we can add a code comment above this line ' something edge case for video audio sample may have different audio formats'

Alex-Wengg commented on Sources/FluidAudio/Shared/AudioStream.swift:
```diff
@@ -0,0 +1,432 @@
+import Foundation
+
+/// A sliding window buffer for a real-time audio stream.
+public class AudioStream {
```
thoughts on making this an actor class to make it thread safe ?

SGD2718 replied:
If an audio stream is being passed around across different tasks instead of having it's chunks being copied and passed around, then that probably means it was designed poorly. It would make more sense to have a central external manager that grabs the chunks and deploys them across concurrent branches

Alex-Wengg replied:
is this class design for 
- Single consumer
- Multiple read-only consumers
- Multiple read-write consumers

Alex-Wengg replied:
alright lmk when you made the changes. i will approve right after

SGD2718 commented:
I also made it possible to silence the AudioConverter's debug logs because they were getting extremely annoying when I was trying to do real-time diarization tests
